
# Ejemplo: matriz de covarianzas

Tomemos las observaciones:

| i | Presión (atm) | Temperatura (K) |
|---|---:|---:|
| 1 | 0.98 | 267.52 |
| 2 | 1.01 | 275.71 |
| 3 | 0.92 | 251.14 |
| 4 | 1.05 | 286.63 |
| 5 | 0.99 | 270.25 |


In [4]:
import numpy as np

def f(P,V,n,R):
    R = 0.082057 # L·atm/(K·mol)
    T = P*V/(n*R)
    return T

P = [0.98, 1.01, 0.92, 1.05, 0.99]
n = 1.0
V = 22.4
T = [ f(P,V,n,0) for P in P]

X = np.column_stack((P,np.round(T,2)))

print(X)

[[  0.98 267.52]
 [  1.01 275.71]
 [  0.92 251.14]
 [  1.05 286.63]
 [  0.99 270.25]]


Tenemos $N=5$ observaciones de $p=2$ variables, en notación del ESL:

$$\mathbf{X} = \begin{pmatrix}
0.98 & 267.52 \\
1.01 & 275.71 \\
0.92 & 251.14 \\
1.05 & 286.63 \\
0.99 & 270.25
\end{pmatrix}$$

La covarianza indica el grado de **variación conjunta** de dos variables aleatorias respecto a sus medias. Es el dato básico para determinar si existe una **dependencia entre ambas variables**.

La covarianza entre dos variables se calcula como:

$$\text{Cov}(X_1, X_2) = \frac{1}{N-1} \sum_{i=1}^{N} (X_{i1} - \bar{X}_1)(X_{i2} - \bar{X}_2)$$

de donde se ve que $\text{Cov}(X_1, X_2) = \text{Cov}(X_2, X_1)$

Siendo la covarianza de una variable consigo misma, la varianza:

$$\text{Cov}(X_1,X_1) = \frac{1}{N-1} \sum_{i=1}^{n} (X_{i1} - \bar{X_1})^2 = \text{Var}(X_1)$$

Matricialmente, la matriz de covarianzas se define como:
$$\mathbf{\Sigma} = \frac{1}{N-1} \mathbf{X_c}^T \mathbf{X_c}$$

donde:

- $\mathbf{\Sigma}_{ij} = \text{Cov}(X_i, X_j)$
- $\mathbf{X_c}$ es la matriz de datos centrada, es decir, cada columna de $\mathbf{X}$ menos su media.

In [5]:
Pm = np.mean(X[:,0])
Tm = np.mean(X[:,1])

Pc = X[:,0] - Pm
Tc = X[:,1] - Tm

Xc = np.column_stack((Pc, Tc))

print(Xc)

[[-1.000e-02 -2.730e+00]
 [ 2.000e-02  5.460e+00]
 [-7.000e-02 -1.911e+01]
 [ 6.000e-02  1.638e+01]
 [ 0.000e+00  0.000e+00]]


In [6]:
#Calculo matriz de covarianzas

S = np.dot(Xc.T, Xc) / (len(Xc)-1)

print(S)

[[2.2500000e-03 6.1425000e-01]
 [6.1425000e-01 1.6769025e+02]]


In [7]:
# Determinante de la matriz de covarianzas

det_S = np.linalg.det(S)

print(det_S)

-6.819544928760015e-17


Teniendo la forma:
$$\mathbf{\Sigma} = \begin{pmatrix}
\text{Var}(X_1) & \text{Cov}(X_1, X_2) \\
\text{Cov}(X_2, X_1) & \text{Var}( X_2) 
\end{pmatrix}$$

En este caso queda entonces:

$$\mathbf{\Sigma} = \begin{pmatrix}
2.25 \times 10^{-3} & 6.14 \times 10^{-1} \\
6.14 \times 10^{-1} & 1.68 \times 10^{2}
\end{pmatrix}$$

Cumpliéndose naturalmente que $\mathbf{\Sigma} = \mathbf{\Sigma}^T$, con lo cual es **simétrica**.

Para ver si una matriz es **definida positiva** podemos:

- calcular los **autovalores** de la matriz, ya que si todos son positivos, entonces la matriz es **definida positiva**, y si todos son no negativos, entonces es **semidefinida positiva**;



- calcular el **determinante** de la matriz y de sus submatrices principales, ya que si todos son positivos, entonces la matriz es **definida positiva**, y si todos son no negativos, entonces es **semidefinida positiva**; o

- calcular el **producto** $\mathbf{x}^T \mathbf{\Sigma} \mathbf{x}$ y analizar su signo, ya que si $\mathbf{x}^T \mathbf{\Sigma} \mathbf{x} > 0$ para todo $\mathbf{x} \neq 0$, entonces $\mathbf{\Sigma}$ es **definida positiva**, y si $\mathbf{x}^T \mathbf{\Sigma} \mathbf{x} \geq 0$ para todo $\mathbf{x} \neq 0$, entonces $\mathbf{\Sigma}$ es **semidefinida positiva**.

Probemos este último camino...

$$\mathbf{x}^T \mathbf{\Sigma} \mathbf{x} = \begin{pmatrix} x_1 & x_2 \end{pmatrix} \begin{pmatrix}
2.25 \times 10^{-3} & 6.14 \times 10^{-1} \\
6.14 \times 10^{-1} & 1.68 \times 10^{2}
\end{pmatrix} \begin{pmatrix} x_1 \\ x_2 \end{pmatrix} $$



$$= 2.25 \times 10^{-3} x_1^2 + 2 \cdot 6.14 \times 10^{-1} x_1 x_2 + 1.68 \times 10^{2} x_2^2$$

quedan productos cruzados y no es fácil ver el signo de la expresión...

Se puede ver así:

Reemplazando $\mathbf{S}$ en $\mathbf{x}^T \mathbf{\Sigma} \mathbf{x}$:

$$\mathbf{x}^T \mathbf{\Sigma} \mathbf{x} = \frac{1}{N-1} \mathbf{x}^T \mathbf{X_c}^T \mathbf{X_c} \mathbf{x}$$

$$= \frac{1}{N-1} (\mathbf{X_c}\mathbf{x})^T (\mathbf{X_c}\mathbf{x}) $$

$$= \frac{1}{N-1} ||\mathbf{X_c}\mathbf{x}||^2$$

que nunca puede ser negativo, y es cero solo si $\mathbf{X_c}\mathbf{x} = 0$.

En este caso particular, $\text{det}(\mathbf{\Sigma}) = 0$, dada la dependencia lineal entre las variables, y por lo tanto $\mathbf{\Sigma}$ es **semidefinida positiva**.

¿Qué pasa si agregamos una variable más?

| i | Presión (atm) | Temperatura (K) | Tiempo de vida media (años)
|---|---:|---:|---:
| 1 | 0.98 | 267.52 | 10.76
| 2 | 1.01 | 275.71 | 12.3
| 3 | 0.92 | 251.14 | 10,403
| 4 | 1.05 | 286.63 | 1.74 $\times 10^{-6}$
| 5 | 0.99 | 270.25 | 301,000

¿Cómo se modifica la matriz de covarianzas? ¿Sigue siendo semidefinida positiva?


Muchas veces resulta útil mirar la **matriz de correlaciones**, que no es más que la matriz de covarianzas de las variables normalizadas, es decir, de las variables **centradas y escaladas por su desviación estándar**:

$$\mathbf{Z} = \begin{pmatrix}
\frac{X_{11} - \bar{X}_1}{s_1} & \frac{X_{12} - \bar{X}_2}{s_2} \\
\frac{X_{21} - \bar{X}_1}{s_1} & \frac{X_{22} - \bar{X}_2}{s_2} \\
\vdots & \vdots \\
\frac{X_{N1} - \bar{X}_1}{s_1} & \frac{X_{N2} - \bar{X}_2}{s_2}
\end{pmatrix}$$

Así la matriz de correlaciones queda:

$$\mathbf{R} = \frac{1}{N-1} \mathbf{Z}^T \mathbf{Z}$$

In [8]:
#Calculo matriz de correlaciones

#Resto la media y divido por la desviación estándar
Z = (X - X.mean(axis=0)) / X.std(axis=0, ddof=1)

R = np.dot(Z.T, Z) / (len(Z)-1)

print(R)

[[1. 1.]
 [1. 1.]]


En el primer caso (para las variables presión y temperatura) la matriz de correlaciones es:

$$\mathbf{R} = \begin{pmatrix}
1 & 1 \\
1 & 1
\end{pmatrix}$$

donde se ve que la correlación entre presión y temperatura es máxima, dado que la temperatura depende linealmente de la presión.

¿Qué quiere decir esto sobre el origen de los datos?

¿Es también la matriz de correlaciones **semidefinida positiva**?